server.py

```diff
@@ -36,17 +36,25 @@ def index():
 ######################################################################
 @app.route('/products', methods=['GET'])
 def list_products():
-    category        = request.args.get('category')
-    discontinued    = str2bool(request.args.get('discontinued'))
-    results     = []
-    for key, value in products.iteritems():
-        if matchesClause(category,      value['category']) and \
-           matchesClause(discontinued,  value['discontinued']):
-             results.append(products[key])
-    return reply(results, HTTP_200_OK)
+	category=request.args.get('category')
+	discontinued = str2bool(request.args.get('discontinued'))
+	search_term=(request.args.get('search')).lower()
```

Comment:
If the API user does not pass a 'search' parameter, I would expect `request.args.get('search')` to return None; in such a case, calling `lower()` on `None` will throw some error like: 
```AttributeError: 'NoneType' object has no attribute 'lower'```

As mentioned below, I believe a helper method (e.g. `matches_search_term(search_term, value)`) would greatly simplify this logic, and inside this method you could first check if the `search` parameter is `None`, and if not, set the value to lower-case. This is a ideal example of encapsulation.

So the operation of setting the search term to lower-case should be performed within the helper method.